tests/test_pipeline_async.py

```diff
@@ -56,3 +96,117 @@ async def test_server_cursor(aconn):
     async with aconn.pipeline():
         with pytest.raises(psycopg.NotSupportedError):
             await cur.execute("select 1")
+
+
+async def test_pipeline_processed_at_exit(aconn):
+    async with aconn.cursor() as cur, aconn.pipeline() as pipeline:
+        await cur.execute("select 1")
+        await pipeline.sync()
+
+        # PQsendQuery[BEGIN], PQsendQuery, PQpipelineSync
+        assert len(pipeline) == 3
+
+    assert len(pipeline) == 0
+    assert await cur.fetchone() == (1,)
+
+
+async def test_pipeline(aconn):
+    async with aconn.pipeline() as pipeline:
+        c1 = aconn.cursor()
+        c2 = aconn.cursor()
+        await c1.execute("select 1")
+        await c2.execute("select 2")
+        await pipeline.sync()
+
+        # PQsendQuery[BEGIN], PQsendQuery(2), PQpipelineSync
+        assert len(pipeline) == 4
+
+        (r1,) = await c1.fetchone()
+        assert r1 == 1
+        assert len(pipeline) == 0
+
+    (r2,) = await c2.fetchone()
+    assert r2 == 2
+
+
+async def test_autocommit(aconn):
+    await aconn.set_autocommit(True)
+    async with aconn.pipeline() as pipeline, aconn.cursor() as c:
+        await c.execute("select 1")
+        await pipeline.sync()
+
+        # PQsendQuery, PQpipelineSync
+        assert len(pipeline) == 2
+
+        (r,) = await c.fetchone()
+        assert r == 1
+
+
+async def test_pipeline_aborted(aconn):
+    await aconn.set_autocommit(True)
+    async with aconn.pipeline() as pipeline:
+        c1 = await aconn.execute("select 1")
+        await pipeline.sync()
+        c2 = await aconn.execute("select * from doesnotexist")
+        c3 = await aconn.execute("select 'aborted'")
+        # Here we can sometimes see that pipeline.status is ABORTED, but this
+        # depends on whether results got fetched by previous execute() and
+        # pipeline communication.
+        await pipeline.sync()
+        c4 = await aconn.execute("select 2")
+        await pipeline.sync()
+
+        (r,) = await c1.fetchone()
+        assert r == 1
+
+        with pytest.raises(UndefinedTable):
+            await c2.fetchone()
+
+        with pytest.raises(psycopg.OperationalError, match="pipeline aborted"):
```

Comment:
This can probably undergo localization: it would be safer to check the SQLSTATE of the exception

Reply:
The message comes from us :)